README.md
# Rest Client

![Build](https://github.com/danblack/idea-rest-client/workflows/Build/badge.svg)
[![Version](https://img.shields.io/jetbrains/plugin/v/9232-http-editor-client.svg)](https://plugins.jetbrains.com/plugin/9232-http-editor-client)
[![Downloads](https://img.shields.io/jetbrains/plugin/d/9232-http-editor-client.svg)](https://plugins.jetbrains.com/plugin/9232-http-editor-client)

<!-- Plugin description -->
Allows making http requests using IDEA text editor

## Features
- Supports GET, POST, PUT, DELETE, PATCH, HEAD & OPTIONS requests
- Multiline request parameters
- Response auto format (based on response Content-Type header)
- Customized colors and fonts
- Customized keyboard shortcuts
- Comments

![demo](doc/demo.png)

## YouTube demos 

* [Simple demo](https://www.youtube.com/watch?v=AliJaGmXlxc)

```bash
http://localhost:8080/test
@Content-type: application/json
%%%

# Duration: 539 ms
# URL: http://localhost:8080/test

# HTTP/1.1 200

@Content-Type: application/json;charset=UTF-8
@Transfer-Encoding: chunked
@Date: Fri, 04 Nov 2016 08:51:00 GMT

{
  "id": 2,
  "name": "resp name",
  "item": [
    {
      "id": 21,
      "name": "item 1",
      "price": 1
    },
    {
      "id": 22,
      "name": "item 2",
      "price": 10
    }
  ]
}
```

* [POST demo](https://www.youtube.com/watch?v=s4oVpfAJHFM)

```bash
POST
http://localhost:8080/mirror-post
@Content-Type: application/json
{"name":"JsonDan"}
%%%

# Duration: 501 ms
# URL: http://localhost:8080/mirror-post

# HTTP/1.1 200

@Content-Type: application/json;charset=UTF-8
@Transfer-Encoding: chunked
@Date: Fri, 04 Nov 2016 11:59:34 GMT

{
  "name": "JsonDan"
}
```

* [Request params demo](https://www.youtube.com/watch?v=Gt8OnWPiJUY)

```bash
http://localhost:8080/mirror-params?one=1
#&two=2&three=3
#&four=4
&five=5
%%%

# Duration: 666 ms
# URL: http://localhost:8080/mirror-params?one=1&five=5

# HTTP/1.1 200

@Content-Type: text/plain;charset=ISO-8859-1
@Content-Length: 14
@Date: Fri, 04 Nov 2016 12:19:52 GMT

one: 1
five: 5
```

<!-- Plugin description end -->

## Installation

- Using IDE built-in plugin system:
  
  <kbd>Preferences</kbd> > <kbd>Plugins</kbd> > <kbd>Marketplace</kbd> > <kbd>Search for "%NAME%"</kbd> >
  <kbd>Install Plugin</kbd>
  
- Manually:

  Download the [latest release](https://github.com/%REPOSITORY%/releases/latest) and install it manually using
  <kbd>Preferences</kbd> > <kbd>Plugins</kbd> > <kbd>⚙️</kbd> > <kbd>Install plugin from disk...</kbd>


CHANGELOG.md
<!-- Keep a Changelog guide -> https://keepachangelog.com -->

# Changelog

## [Unreleased]
### Changed
- Migrated the plugin to gradle

### Fixed
- Fixed issues with deprecated APIs

## [0.1.1]
### Added
- Basic support


src/ru/basecode/ide/rest/plugin/RestLexer.java
/* The following code was generated by JFlex 1.4.3 on 04.04.16 22:28 */

package ru.basecode.ide.rest.plugin;

import com.intellij.lexer.FlexLexer;
import com.intellij.psi.tree.IElementType;
import ru.basecode.ide.rest.plugin.psi.RestTypes;
import com.intellij.psi.TokenType;


/**
 * This class is a scanner generated by 
 * <a href="http://www.jflex.de/">JFlex</a> 1.4.3
 * on 04.04.16 22:28 from the specification file
 * <tt>/Users/zoom/Development/ide/projects/idea-rest-client/src/ru/basecode/ide/rest/plugin/Rest.flex</tt>
 */
class RestLexer implements FlexLexer {
  /** initial size of the lookahead buffer */
  private static final int ZZ_BUFFERSIZE = 16384;

  /** lexical states */
  public static final int S_RESP_HEADER = 12;
  public static final int S_PARAM = 8;
  public static final int S_RESP_BODY = 14;
  public static final int S_REQ_HEADER = 10;
  public static final int S_METHOD = 2;
  public static final int S_URL = 6;
  public static final int YYINITIAL = 0;
  public static final int S_REQ_BODY = 16;
  public static final int S_OPTION = 4;

  /**
   * ZZ_LEXSTATE[l] is the state in the DFA for the lexical state l
   * ZZ_LEXSTATE[l+1] is the state in the DFA for the lexical state l
   *                  at the beginning of a line
   * l is of the form l = 2*k, k a non negative integer
   */
  private static final int ZZ_LEXSTATE[] = { 
     0,  0,  1,  1,  2,  2,  3,  3,  4,  4,  5,  5,  6,  6,  7,  7, 
     8, 8
  };

  /** 
   * Translates characters to character classes
   */
  private static final String ZZ_CMAP_PACKED = 
    "\11\0\1\31\1\27\1\0\1\31\1\30\22\0\1\31\2\0\1\20"+
    "\1\0\1\32\1\17\6\0\1\15\1\0\1\26\12\0\1\25\5\0"+
    "\1\16\1\10\1\0\1\11\1\13\1\2\1\0\1\1\1\12\3\0"+
    "\1\14\2\0\1\5\1\4\2\0\1\6\1\3\1\7\22\0\1\21"+
    "\7\0\1\23\2\0\1\24\1\22\uff8b\0";

  /** 
   * Translates characters to character classes
   */
  private static final char [] ZZ_CMAP = zzUnpackCMap(ZZ_CMAP_PACKED);

  /** 
   * Translates DFA states to action switch labels.
   */
  private static final int [] ZZ_ACTION = zzUnpackAction();

  private static final String ZZ_ACTION_PACKED_0 =
    "\11\0\5\1\1\2\1\1\2\3\1\4\1\1\1\5"+
    "\1\6\1\7\1\4\1\5\1\10\1\11\1\12\1\13"+
    "\4\0\1\14\1\0\1\15\1\0\1\5\1\16\4\0"+
    "\1\17\3\0\1\17\4\0\1\20";

  private static int [] zzUnpackAction() {
    int [] result = new int[53];
    int offset = 0;
    offset = zzUnpackAction(ZZ_ACTION_PACKED_0, offset, result);
    return result;
  }

  private static int zzUnpackAction(String packed, int offset, int [] result) {
    int i = 0;       /* index in packed string  */
    int j = offset;  /* index in unpacked array */
    int l = packed.length();
    while (i < l) {
      int count = packed.charAt(i++);
      int value = packed.charAt(i++);
      do result[j++] = value; while (--count > 0);
    }
    return j;
  }


  /** 
   * Translates a state to a row index in the transition table
   */
  private static final int [] ZZ_ROWMAP = zzUnpackRowMap();

  private static final String ZZ_ROWMAP_PACKED_0 =
    "\0\0\0\33\0\66\0\121\0\154\0\207\0\242\0\275"+
    "\0\330\0\363\0\u010e\0\u0129\0\u0144\0\u015f\0\u017a\0\u0195"+
    "\0\363\0\u01b0\0\u01cb\0\u01e6\0\u0201\0\u021c\0\u0237\0\u0252"+
    "\0\u026d\0\u0288\0\u02a3\0\u02be\0\u02d9\0\u02f4\0\u030f\0\u032a"+
    "\0\u0345\0\u0360\0\u037b\0\u0396\0\u03b1\0\u03cc\0\363\0\u03e7"+
    "\0\u0402\0\u041d\0\u0438\0\u0453\0\u046e\0\u0489\0\u04a4\0\u04bf"+
    "\0\u04da\0\u04f5\0\u0510\0\u052b\0\u0546";

  private static int [] zzUnpackRowMap() {
    int [] result = new int[53];
    int offset = 0;
    offset = zzUnpackRowMap(ZZ_ROWMAP_PACKED_0, offset, result);
    return result;
  }

  private static int zzUnpackRowMap(String packed, int offset, int [] result) {
    int i = 0;  /* index in packed string  */
    int j = offset;  /* index in unpacked array */
    int l = packed.length();
    while (i < l) {
      int high = packed.charAt(i++) << 16;
      result[j++] = high | packed.charAt(i++);
    }
    return j;
  }

  /** 
   * The transition table of the DFA
   */
  private static final int [] ZZ_TRANS = zzUnpackTrans();

  private static final String ZZ_TRANS_PACKED_0 =
    "\1\12\1\13\2\12\1\14\6\12\1\15\1\12\1\16"+
    "\2\12\1\17\1\20\5\12\1\21\1\22\1\23\21\12"+
    "\1\17\1\20\5\12\1\21\1\22\1\23\2\12\1\13"+
    "\2\12\1\14\6\12\1\15\1\12\1\24\2\12\1\17"+
    "\1\20\5\12\1\21\1\22\1\23\1\12\16\25\1\26"+
    "\1\27\1\17\6\25\1\21\1\22\1\30\1\31\16\25"+
    "\1\26\1\32\1\17\6\25\1\21\1\22\1\30\1\31"+
    "\16\25\1\33\1\25\1\17\6\25\1\21\1\22\1\30"+
    "\1\31\16\34\1\33\1\34\1\17\6\34\1\21\1\22"+
    "\1\23\1\34\27\35\1\21\1\22\1\23\1\35\27\25"+
    "\1\21\1\22\1\30\1\31\35\0\1\36\35\0\1\37"+
    "\1\0\1\36\1\40\24\0\1\41\45\0\1\42\15\0"+
    "\27\17\2\0\2\17\22\0\1\43\37\0\1\21\34\0"+
    "\1\23\16\0\1\44\15\0\27\25\2\0\2\25\27\26"+
    "\2\0\2\26\27\27\2\0\2\27\31\0\1\30\1\45"+
    "\27\25\2\0\1\25\1\46\27\32\2\0\2\32\27\33"+
    "\2\0\2\33\27\34\2\0\2\34\27\35\2\0\2\35"+
    "\3\0\1\47\35\0\1\36\27\0\1\50\43\0\1\51"+
    "\16\0\27\42\2\0\2\42\22\0\1\52\10\0\27\44"+
    "\2\0\2\44\32\0\1\53\27\25\2\0\1\25\1\54"+
    "\11\0\1\55\23\0\1\56\53\0\1\57\41\0\1\60"+
    "\27\25\2\0\1\54\1\25\12\0\1\47\23\0\1\61"+
    "\53\0\1\62\1\63\36\0\1\60\3\0\1\47\55\0"+
    "\1\63\33\0\1\64\32\0\1\65\4\0\27\65\2\0"+
    "\2\65";

  private static int [] zzUnpackTrans() {
    int [] result = new int[1377];
    int offset = 0;
    offset = zzUnpackTrans(ZZ_TRANS_PACKED_0, offset, result);
    return result;
  }

  private static int zzUnpackTrans(String packed, int offset, int [] result) {
    int i = 0;       /* index in packed string  */
    int j = offset;  /* index in unpacked array */
    int l = packed.length();
    while (i < l) {
      int count = packed.charAt(i++);
      int value = packed.charAt(i++);
      value--;
      do result[j++] = value; while (--count > 0);
    }
    return j;
  }


  /* error codes */
  private static final int ZZ_UNKNOWN_ERROR = 0;
  private static final int ZZ_NO_MATCH = 1;
  private static final int ZZ_PUSHBACK_2BIG = 2;
  private static final char[] EMPTY_BUFFER = new char[0];
  private static final int YYEOF = -1;
  private static java.io.Reader zzReader = null; // Fake

  /* error messages for the codes above */
  private static final String ZZ_ERROR_MSG[] = {
    "Unkown internal scanner error",
    "Error: could not match input",
    "Error: pushback value was too large"
  };

  /**
   * ZZ_ATTRIBUTE[aState] contains the attributes of state <code>aState</code>
   */
  private static final int [] ZZ_ATTRIBUTE = zzUnpackAttribute();

  private static final String ZZ_ATTRIBUTE_PACKED_0 =
    "\11\0\1\11\6\1\1\11\14\1\4\0\1\1\1\0"+
    "\1\1\1\0\1\1\1\11\4\0\1\1\3\0\1\1"+
    "\4\0\1\1";

  private static int [] zzUnpackAttribute() {
    int [] result = new int[53];
    int offset = 0;
    offset = zzUnpackAttribute(ZZ_ATTRIBUTE_PACKED_0, offset, result);
    return result;
  }

  private static int zzUnpackAttribute(String packed, int offset, int [] result) {
    int i = 0;       /* index in packed string  */
    int j = offset;  /* index in unpacked array */
    int l = packed.length();
    while (i < l) {
      int count = packed.charAt(i++);
      int value = packed.charAt(i++);
      do result[j++] = value; while (--count > 0);
    }
    return j;
  }

  /** the current state of the DFA */
  private int zzState;

  /** the current lexical state */
  private int zzLexicalState = YYINITIAL;

  /** this buffer contains the current text to be matched and is
      the source of the yytext() string */
  private CharSequence zzBuffer = "";

  /** this buffer may contains the current text array to be matched when it is cheap to acquire it */
  private char[] zzBufferArray;

  /** the textposition at the last accepting state */
  private int zzMarkedPos;

  /** the textposition at the last state to be included in yytext */
  private int zzPushbackPos;

  /** the current text position in the buffer */
  private int zzCurrentPos;

  /** startRead marks the beginning of the yytext() string in the buffer */
  private int zzStartRead;

  /** endRead marks the last character in the buffer, that has been read
      from input */
  private int zzEndRead;

  /**
   * zzAtBOL == true <=> the scanner is currently at the beginning of a line
   */
  private boolean zzAtBOL = true;

  /** zzAtEOF == true <=> the scanner is at the EOF */
  private boolean zzAtEOF;

  /** denotes if the user-EOF-code has already been executed */
  private boolean zzEOFDone;


  /**
   * Creates a new scanner
   *
   * @param   in  the java.io.Reader to read input from.
   */
  RestLexer(java.io.Reader in) {
    this.zzReader = in;
  }


  /** 
   * Unpacks the compressed character translation table.
   *
   * @param packed   the packed character translation table
   * @return         the unpacked character translation table
   */
  private static char [] zzUnpackCMap(String packed) {
    char [] map = new char[0x10000];
    int i = 0;  /* index in packed string  */
    int j = 0;  /* index in unpacked array */
    while (i < 92) {
      int  count = packed.charAt(i++);
      char value = packed.charAt(i++);
      do map[j++] = value; while (--count > 0);
    }
    return map;
  }

  public final int getTokenStart(){
    return zzStartRead;
  }

  public final int getTokenEnd(){
    return getTokenStart() + yylength();
  }

  public void reset(CharSequence buffer, int start, int end,int initialState){
    zzBuffer = buffer;
    zzBufferArray = com.intellij.util.text.CharArrayUtil.fromSequenceWithoutCopying(buffer);
    zzCurrentPos = zzMarkedPos = zzStartRead = start;
    zzPushbackPos = 0;
    zzAtEOF  = false;
    zzAtBOL = true;
    zzEndRead = end;
    yybegin(initialState);
  }

  /**
   * Refills the input buffer.
   *
   * @return      <code>false</code>, iff there was new input.
   *
   * @exception   java.io.IOException  if any I/O-Error occurs
   */
  private boolean zzRefill() throws java.io.IOException {
    return true;
  }


  /**
   * Returns the current lexical state.
   */
  public final int yystate() {
    return zzLexicalState;
  }


  /**
   * Enters a new lexical state
   *
   * @param newState the new lexical state
   */
  public final void yybegin(int newState) {
    zzLexicalState = newState;
  }


  /**
   * Returns the text matched by the current regular expression.
   */
  public final CharSequence yytext() {
    return zzBuffer.subSequence(zzStartRead, zzMarkedPos);
  }


  /**
   * Returns the character at position <tt>pos</tt> from the
   * matched text.
   *
   * It is equivalent to yytext().charAt(pos), but faster
   *
   * @param pos the position of the character to fetch.
   *            A value from 0 to yylength()-1.
   *
   * @return the character at position pos
   */
  public final char yycharat(int pos) {
    return zzBufferArray != null ? zzBufferArray[zzStartRead+pos]:zzBuffer.charAt(zzStartRead+pos);
  }


  /**
   * Returns the length of the matched text region.
   */
  public final int yylength() {
    return zzMarkedPos-zzStartRead;
  }


  /**
   * Reports an error that occured while scanning.
   *
   * In a wellformed scanner (no or only correct usage of
   * yypushback(int) and a match-all fallback rule) this method
   * will only be called with things that "Can't Possibly Happen".
   * If this method is called, something is seriously wrong
   * (e.g. a JFlex bug producing a faulty scanner etc.).
   *
   * Usual syntax/scanner level error handling should be done
   * in error fallback rules.
   *
   * @param   errorCode  the code of the errormessage to display
   */
  private void zzScanError(int errorCode) {
    String message;
    try {
      message = ZZ_ERROR_MSG[errorCode];
    }
    catch (ArrayIndexOutOfBoundsException e) {
      message = ZZ_ERROR_MSG[ZZ_UNKNOWN_ERROR];
    }

    throw new Error(message);
  }


  /**
   * Pushes the specified amount of characters back into the input stream.
   *
   * They will be read again by then next call of the scanning method
   *
   * @param number  the number of characters to be read again.
   *                This number must not be greater than yylength()!
   */
  public void yypushback(int number)  {
    if ( number > yylength() )
      zzScanError(ZZ_PUSHBACK_2BIG);

    zzMarkedPos -= number;
  }


  /**
   * Contains user EOF-code, which will be executed exactly once,
   * when the end of file is reached
   */
  private void zzDoEOF() {
    if (!zzEOFDone) {
      zzEOFDone = true;
    
    }
  }


  /**
   * Resumes scanning until the next regular expression is matched,
   * the end of input is encountered or an I/O-Error occurs.
   *
   * @return      the next token
   * @exception   java.io.IOException  if any I/O-Error occurs
   */
  public IElementType advance() throws java.io.IOException {
    int zzInput;
    int zzAction;

    // cached fields:
    int zzCurrentPosL;
    int zzMarkedPosL;
    int zzEndReadL = zzEndRead;
    CharSequence zzBufferL = zzBuffer;
    char[] zzBufferArrayL = zzBufferArray;
    char [] zzCMapL = ZZ_CMAP;

    int [] zzTransL = ZZ_TRANS;
    int [] zzRowMapL = ZZ_ROWMAP;
    int [] zzAttrL = ZZ_ATTRIBUTE;

    while (true) {
      zzMarkedPosL = zzMarkedPos;

      zzAction = -1;

      zzCurrentPosL = zzCurrentPos = zzStartRead = zzMarkedPosL;

      zzState = ZZ_LEXSTATE[zzLexicalState];


      zzForAction: {
        while (true) {

          if (zzCurrentPosL < zzEndReadL)
            zzInput = (zzBufferArrayL != null ? zzBufferArrayL[zzCurrentPosL++] : zzBufferL.charAt(zzCurrentPosL++));
          else if (zzAtEOF) {
            zzInput = YYEOF;
            break zzForAction;
          }
          else {
            // store back cached positions
            zzCurrentPos  = zzCurrentPosL;
            zzMarkedPos   = zzMarkedPosL;
            boolean eof = zzRefill();
            // get translated positions and possibly new buffer
            zzCurrentPosL  = zzCurrentPos;
            zzMarkedPosL   = zzMarkedPos;
            zzBufferL      = zzBuffer;
            zzEndReadL     = zzEndRead;
            if (eof) {
              zzInput = YYEOF;
              break zzForAction;
            }
            else {
              zzInput = (zzBufferArrayL != null ? zzBufferArrayL[zzCurrentPosL++] : zzBufferL.charAt(zzCurrentPosL++));
            }
          }
          int zzNext = zzTransL[ zzRowMapL[zzState] + zzCMapL[zzInput] ];
          if (zzNext == -1) break zzForAction;
          zzState = zzNext;

          int zzAttributes = zzAttrL[zzState];
          if ( (zzAttributes & 1) == 1 ) {
            zzAction = zzState;
            zzMarkedPosL = zzCurrentPosL;
            if ( (zzAttributes & 8) == 8 ) break zzForAction;
          }

        }
      }

      // store back cached position
      zzMarkedPos = zzMarkedPosL;

      switch (zzAction < 0 ? zzAction : ZZ_ACTION[zzAction]) {
        case 7: 
          { yybegin(S_PARAM); return RestTypes.PARAM;
          }
        case 17: break;
        case 4: 
          { return RestTypes.WHITE_SPACE;
          }
        case 18: break;
        case 11: 
          { return RestTypes.RESPONSE_BODY_LINE;
          }
        case 19: break;
        case 9: 
          { return RestTypes.HEADER;
          }
        case 20: break;
        case 14: 
          { yybegin(S_METHOD); return RestTypes.METHOD;
          }
        case 21: break;
        case 2: 
          { return RestTypes.COMMENT;
          }
        case 22: break;
        case 1: 
          { return RestTypes.BAD_CHARACTER;
          }
        case 23: break;
        case 15: 
          { yybegin(S_RESP_HEADER); return RestTypes.SEPARATOR;
          }
        case 24: break;
        case 16: 
          { yybegin(S_URL); return RestTypes.URL;
          }
        case 25: break;
        case 3: 
          { return RestTypes.CRLF;
          }
        case 26: break;
        case 10: 
          { yybegin(S_RESP_BODY); return RestTypes.RESPONSE_BODY_LINE;
          }
        case 27: break;
        case 6: 
          { yybegin(S_REQ_HEADER); return RestTypes.HEADER;
          }
        case 28: break;
        case 12: 
          { yybegin(S_OPTION); return RestTypes.OPTION;
          }
        case 29: break;
        case 8: 
          { return RestTypes.PARAM;
          }
        case 30: break;
        case 13: 
          { return RestTypes.OPTION;
          }
        case 31: break;
        case 5: 
          { yybegin(S_REQ_BODY); return RestTypes.REQUEST_BODY_LINE;
          }
        case 32: break;
        default:
          if (zzInput == YYEOF && zzStartRead == zzCurrentPos) {
            zzAtEOF = true;
            zzDoEOF();
            return null;
          }
          else {
            zzScanError(ZZ_NO_MATCH);
          }
      }
    }
  }


}
